src/buffer/out/textBuffer.cpp

```diff
@@ -376,6 +376,24 @@ bool TextBuffer::_PrepareForDoubleByteSequence(const DbcsAttribute dbcsAttribute
     return fSuccess;
 }
 
+void TextBuffer::ConsumeGrapheme(std::wstring_view& chars) noexcept
+{
+    // This function is supposed to mirror the behavior of ROW::Write, when it reads characters off of `chars`.
+    // (I know that a UTF-16 code point is not a grapheme, but that's what we're working towards.)
+    chars = til::utf16_pop(chars);
+}
+
+void TextBuffer::Write(til::CoordType row, bool wrapAtEOL, const TextAttribute& attributes, RowWriteState& state)
+{
+    auto& r = GetRowByOffset(row);
+
+    r.ReplaceText(state);
+    r.ReplaceAttributes(state.columnBegin, state.columnEnd, attributes);
+    r.SetWrapForced(wrapAtEOL && state.columnEndDirty >= r.size());

```

Comment:
This is a bit of an edge case, but if you've got one column free at the end of the line, and you're trying to write out a wide glyph, you won't replace any attributes because `columnEnd` will equal `columnBegin`. But this is one case when I think that `ReplaceAttributes` call should possibly be using `columnEndDirty` (not sure of the exact logic, but something along those lines). Because even though you haven't stored anything in that last column, it has effectively been overwritten and should really take the active attribute color.

For example, in the test case below I've got a couple of wide glyphs written with a default background color, that wrap over the end of a line that was originally red. It feels wrong that you appear to end up with a red space between the characters.

    printf "\e[41m\e[K\e[m\e[999C\e[2Dこん\n"

Although if this is a pain to get right, I don't think it's a big deal.

Reply:
I also wanted to note that the `SetWrapForced` call is probably wrong. Writing to the last column shouldn't treat the row as wrapped. That should only happen once you output an additional character that forces the cursor to move to the next line. But this is an existing bug (#3088), so I wouldn't necessarily expect it to be addressed in this PR.

Reply:
BTW I believe this line has another bug: For rows with a double-width line-rendition, we need to check for `state.columnEndDirty >= r.size() / 2`. In my latest commit I've renamed the function to `WriteLine`. It now also automatically gets the line width from the row, so we don't have to get and update it in `AdaptDispatch::_WriteToBuffer` anymore.

I believe the `SetWrapForced` bug you mentioned can be fixed by doing this:
```cpp
r.SetWrapForced(wrapAtEOL && state.columnEnd >= state.columnLimit && !state.text.empty());
```

In other words: It only wraps if the new cursor position (`columnEnd`) is past the line width (`columnLimit`) but there's more text to come (`!text.empty()).

However I found that this breaks the current implementation of `TextBuffer::Reflow`, which has this condition for line breaks: https://github.com/microsoft/terminal/blob/b9248fa903eaa8d187c72bf08f8aa17b25f9d6bb/src/buffer/out/textBuffer.cpp#L2486-L2491

My latest commit removes the `iRight < cOldColsTotal` part of the condition, which then fixes the bug as described in #3088 for both widening and narrowing the viewport. I'm not sure if removing that part of the condition is safe however, because even with the code comment above it, I can't quite figure out why the check exists in the first place. Isn't `row.WasWrapForced()` indication enough that a row had an explicit newline? [^1] I might drop this part of the PR before it merges if I continue  feeling that unconfident about it. 😅

[^1]: My understanding so far is: It always wraps as if a `\r\n` was there, _unless_ `WasWrapForced()` is `true`.

Reply:
> In other words: It only wraps if the new cursor position (`columnEnd`) is past the line width (`columnLimit`) but there's more text to come (`!text.empty()`).

I don't think that completely solves it. You can still have a situation where you write to the last column without more text to come, and then later write additional text that produces a wrap. That wrap is triggered by the `delayedEOLWrap` check in `_WriteToBuffer`, so it doesn't pass through this code, and thus the wrap flag would never be set.

I _think_ the correct place to be handling this, at least for VT code, is in the `_WriteToBuffer` implementation. The legacy console code path may be more complicated - I'm guessing that's going to be somewhere in `WriteCharsLegacy`. There could be all sorts of weird edge cases though. The more I think about it, the more I think it should be left for a follow up PR.